big-o/binary-search.md

```diff
@@ -6,12 +6,100 @@
+            high = mid - 1
+        elif array[mid] < value:
+            low = mid + 1
+        else:
+            return mid
+
+        if array[low] == value:
+            return low
+
+    return None
+```
+
+Let's analyze this algorithm by using a loop table with a concrete example of searching for a `value` in an `array` of integers. Imagine a `array` of eight elements `[1, 2, 3, 4, 5, 6, 7, 8]`. The `value` we are searching for is `2`:
+
+| Iteration | `low` | `high` |`low <= high`| `mid` | 
+|--|--|--|--|--|
+|1| `0` | `7` | `True` | `3` | 
+|2| `0` | `2` | `True` | `1` |
+
+On the second iteration, the line `return mid` is executed because `array[1] = 2`
 
-## Implementation
+Notice with each iteration the size of the input involved in the search (`low` to `high`) is halved. So worst-case a list of 8 items would take 3 iterations to find the value. We could double the size of `numbers` to 16 items and it would only take 4 iteration to find the value and for 32 items it would only take 5 iterations. Thus while the function *does* take longer as the input size increases it does not increase very rapidly.
+
+The table below summarizes the number of iterations in the worst case scenario required to find a `value` with the `binary_search` algorithm for list of length `n`.
+
+|   n |   Number of iterations |
+|--- |--- |
+|   8 |   3 |
+|   16 |   4 |
+|   32 |   5 |
+|   64 |   6 |
+
+*Table. Number of iterations for each input size.*
+
+## Big-O Analysis
+
+### Time Complexity
+
+Recall that time complexity is the measurement of how the amount of time an algorithm takes to run as the size of the input changes. Time complexity is usually measured in _number of operations._
+
+The relationship between the number of operations and the input size for `binary_search` described above is a logarithmic relationship. In Big O notation, this is `O(log n)`.
+
+#### Logorithms
+
+**What is a log?**?
+
+A logarithm is a quantity representing the power to which a fixed number (the base) must be raised to produce a given number. 
+
+| Logarithmic Equation | Equivalent Power Equation | How it is Read | 
+|--|--|--|
+| log <sub>2</sub>8 = 3 | 2<sup>3</sup> = 8 | **log base 2 of 8 equals 3** is equivalent to **2 to the power of 3 equals 8** |
+| log <sub>10</sub>10000 = 4 | 10<sup>4</sup> = 10000 | **log base 10 of 10000 equals 4** is equivalent to **10 to the power of 4 equals 10000** |
+
+In a coding problem if we reduce the size of a problem by dividing the remaining input with each iteration we often get a time complexity involving a logarithm.
+
+In general, in a coding problem:
+
+* If the size of the input is *divided* by some value with each iteration, the time complexity involves a logarithm with a base equal to the divisor.
+* By far, the most common logarithmic base is 2 because our algorithms often, like binary search, divide the input size by two with each iteration.
+  * Often we drop the base of the logarithm on the assumption that it is 2.
```

Comment:
This isn't quite right, we don't need to assume, since all logarithms grow within the same constant factors:

```suggestion
  * Often we drop the base of the logarithm since all logarithms can be [converted into base 2 using constants](https://www.khanacademy.org/math/algebra2/x2ec2f6f830c9fb89:logs/x2ec2f6f830c9fb89:change-of-base/a/logarithm-change-of-base-rule-intro).
```